README.md
collectd-rabbitmq
================= 

A collected plugin, written in python, to collect statistics from RabbitMQ. Requires that the rabbitmq management plugin is installed. 

Installation
============

Place the rabbitmq.py file in the plugins directory for collect. 
Each of these statistics have thier own custom entery in collectd's type database. 
To add these types run:

``` cat config/types.db.custom >> /usr/share/collectd/types.db ```


Configuration 
=============

This plugin supports a small amount of configuration options:

* ```Username```: The rabbitmq user. Defaults to ```guest```.

* ```Password```: The rabbitmq user password. Defaults to ```guest```.

* ```Realm```: The http realm for authentication. Defaults to ```RabbitMQ Management```. 

* ```Host```: The hostname that the rabbitmq server running on. Defaults to ```localhost```.

* ```Port```: The port that the rabbitmq server is listening on. Defaults to ```15672```.

* ```PrefixVhost```: If enabled, rabbitmq vhost will be prefixed with the hostname in the metric name. Defaults to ```false```.

* ```Ignore```: The queue to ignore, matching by Regex.  See example.

Example Configuration
=====================

```
LoadPlugin python
<Plugin python>
  ModulePath "/mnt/projects/collectd-rabbitmq/"
  LogTraces true
  Interactive false
  Import rabbitmq
  <Module rabbitmq>
    Username "guest"
    Password "guest"
    Realm "RabbitMQ Management"
    Host "localhost"
    Port "15672"
    PrefixVhost true
	<Ignore "queue">
	  Regex "amq-gen-.*"
	  Regex "tmp-.*"
	</Ignore>
  </Module>
</Plugin>
```

Nodes
=====

For each node the following statistics are gathered:

* disk_free_limit

* fd_total

* fd_used 

* mem_limit

* mem_used

* proc_total 

* proc_used

* processors

* run_queue

* sockets_total

* sockets_used

Queues
=======

For each queue in each vhost the following statistics are gathered:
> NOTE: The ```/``` vhost name is sent as ```default```

* message_stats
    * deliver_get
    * deliver_get_details 
    	* rate 
    * get
    * get_details
        * rate
    * publish
    * publish_details
        * rate
    * redeliver  
    * redeliver_details
        * rate
* messages
* messages_details 
    * rate
* messages_ready
* messages_ready_details
    * rate
* messages_unacknowledged
* messages_unacknowledged_details
  * rate
* memory
* consumers

Exchanges
=========

For each exchange in each vhost the following statistics are gathered: 
> NOTE: The ```/``` vhost name is sent as ```default```

* disk_free 

* disk_free_limit 

* fd_total

* fd_used

* mem_limit 

* mem_used

* proc_total

* proc_used 

* processors

* run_queue

* sockets_total

* sockets_used


Developing
==========

Whenever a new stat is to be added, it must be added to the config/types.db.custom file. 


HISTORY.rst
.. :changelog:

History
-------

0.1.0 (2014-09-18)
---------------------

* First public release.


tests/utils.py
#!/usr/bin/python
# -*- coding: iso-8859-15 -*-

# Copyright (c) 2014 The New York Times Company
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

""" Simple test modules rabbit test module """

import json
import urlparse


def get_message_stats_data(name):
    """
    Returns fake message stats data.
    """
    return dict(name=name, vhost="test_vhost",
                message_stats=dict(
                    publish_in=10,
                    publish_in_details=dict(rate=20.0),
                    publish_out=10,
                    publish_out_details=dict(rate=110.0)
                ))


def create_mock_url_repsonse(url):
    """
    Returns mocked stats data based on URL.
    """
    name = urlparse.urlparse(url).path.split('/')[-1:][0]
    data = get_message_stats_data(name)
    return MockURLResponse(json.dumps(data))


def get_node_stats_data():
    """
    Returns fake node stats data.
    """
    return [{
        "disk_free": 20234559488,
        "disk_free_details": {
            "rate": -27852.799999999999
        },
        "disk_free_limit": 50000000,

        "fd_total": 150000,
        "fd_used": 113,
        "fd_used_details": {
            "rate": 0.20000000000000001
        },
        "mem_limit": 2030287257,
        "mem_used": 108663688,
        "mem_used_details": {
            "rate": -28052.799999999999
        },
        "name": "rabbit@fabrik",
        "net_ticktime": 60,
        "os_pid": "29958",
        "partitions": [],
        "proc_total": 1048576,
        "proc_used": 1626,
        "proc_used_details": {
            "rate": -6.2000000000000002
        },
        "processors": 2,
        "rates_mode": "basic",
        "run_queue": 0,
        "sockets_total": 134908,
        "sockets_used": 94,
        "sockets_used_details": {
            "rate": -0.20000000000000001
        },
        "type": "disc",
        "uptime": 65806573
    }]


def create_mock_node_url_repsonse(url=None):
    """
    Returns mocked node stats data based on URL.
    """
    data = get_node_stats_data()
    return MockURLResponse(json.dumps(data))


class MockURLResponse(object):
    """
    A class to mock URL lib response
    based on https://gist.github.com/puffin/966992
    """

    def __init__(self, resp_data, code=200, msg='OK'):
        self.resp_data = resp_data
        self.code = code
        self.msg = msg
        self.headers = {'content-type': 'text/plain; charset=utf-8'}

    def read(self):
        """
        Implements read function by returning base data.
        """
        return self.resp_data

    def getcode(self):
        """
        Returns HTTP code.
        """
        return self.code


tests/collectd.py
#!/usr/bin/python
# -*- coding: iso-8859-15 -*-

# Copyright (c) 2014 The New York Times Company
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Mock collected object for testing
Based on:
https://github.com/rampage644/blueflood-python/blob/master/test/collectd.py
"""
import logging

# pylint: disable=R0913
# pylint: disable=R0902
# pylint: disable=R0903
# pylint: disable=W0613


class Config(object):
    """
    Fake config oebject.
    """
    def __init__(self, key, values, children=None):
        # also Config instance
        self.parent = None
        # key string
        self.key = key
        # tuple of Config instances, mostly empty
        self.children = children or list()
        for child in self.children:
            child.parent = self
        # tuple of strings, mostly one value
        self.values = values

    def __repr__(self):
        if self.parent:
            prefix = '\n\t'
        else:
            prefix = ''

        return "{0}Key: {1}, Value: {2}, Parent: {3}, Children: {4}".format(
            prefix, self.key, self.values, hex(id(self.parent)), self.children)


class Values(object):
    """
    Fake value object build for testing.
    Attempts to implement:
        https://collectd.org/documentation/manpages/collectd-python.5.shtml
    """
    def __init__(self, host='', plugin='', plugin_instance='', value_type='',
                 type_instance='', time=0, values=None, interval=0):
        self.host = host
        self.plugin = plugin
        self.plugin_instance = plugin_instance
        self.time = time
        self.value_type = value_type
        self.type_instance = type_instance
        self.values = values or tuple()
        self.interval = interval
        self.meta = None

    def dispatch(self, value_type=None, values=None, plugin_instance=None,
                 type_instance=None, plugin=None, host=None, time=None,
                 interval=None):
        """
        Fake dispatch method, does nothing.
        """
        pass

    def write(self, destination=None, value_type=None, values=None,
              plugin_instance=None, type_instance=None, plugin=None, host=None,
              time=None, interval=None):
        """
        Fake dispatch method, does nothing.
        """
        pass


def register_write(func, data):
    """
    Fake write function.
    """
    pass


def register_config(func):
    """
    Fake config function.
    """
    pass


def register_init(func):
    """
    Fake init function.
    """
    pass


def register_read(func):
    """
    Fake read function.
    """
    pass


def register_shutdown(func):
    """
    Fake shutdown function.
    """
    pass


def info(msg):
    """
    Logs as info level.
    """
    logging.info(msg)


def warning(msg):
    """
    Logs as warning level.
    """
    logging.warning(msg)


def error(msg):
    """
    Logs as error level.
    """
    logging.error(msg)


def debug(msg):
    """
    Logs as error level.
    """
    logging.debug(msg)


tests/test_utils.py
#!/usr/bin/python
# -*- coding: iso-8859-15 -*-

# Copyright (c) 2014 The New York Times Company
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

""" Test module for utilitys"""

import logging
import sys
import unittest

from collectd_rabbitmq import utils


class TestFilterDictionary(unittest.TestCase):
    """
    Test class for filter_dictionary method.
    """

    def test_filter_dictionary(self):
        """
        Asserts that dictionary is filtered properly.
        """
        test_dict = dict(one='test1',
                         two='test2',
                         three='test3')
        test_keys = ['one', 'two']
        filtered = utils.filter_dictionary(test_dict, test_keys)
        self.assertNotIn('three', filtered.keys())

    def test_filter_no_keys(self):
        """
        Asserts that dictionary is filtered properly.
        """
        test_dict = dict(one='test1',
                         two='test2',
                         three='test3')
        filtered = utils.filter_dictionary(test_dict, None)
        self.assertEquals(filtered, dict())

    def test_filter_no_dictionary(self):
        """
        Asserts that dictionary is filtered properly.
        """
        test_keys = ['one', 'two']
        filtered = utils.filter_dictionary(None, test_keys)
        self.assertEquals(filtered, dict())


class TestConnectionInfo(unittest.TestCase):
    """
    Test class for collection object.
    """
    def setUp(self):
        self.conn = utils.ConnectionInfo(host="example.com",
                                         port=15672,
                                         scheme="http")

    def test_url_setter(self):
        """
        Assert the the url is parse and stored properly.
        """
        self.conn.url = "https://test.com:2112"
        self.assertEqual(self.conn.host, "test.com")
        self.assertEqual(self.conn.port, 2112)
        self.assertEqual(self.conn.scheme, "https")


class TestConfig(unittest.TestCase):
    """
    Test class for config object.
    """
    def setUp(self):
        self.conn = utils.ConnectionInfo(host="example.com",
                                         port=15672,
                                         scheme="http")
        self.auth = utils.Auth()

    def test_config_url(self):
        """
        Assert that host and port are set from URL.
        """
        conf = utils.Config(None, self.conn)
        self.assertEquals(conf.connection.host, "example.com")
        self.assertEquals(conf.connection.scheme, "http")
        self.assertEquals(conf.connection.port, 15672)

    def test_config_ignored(self):
        """
        Asserts that ignored_data is properly ignored.
        """
        ignored_data = dict(exchange=['a.*', 'b.*'])
        conf = utils.Config(self.auth, self.conn, ignored_data)
        self.assertTrue(conf.is_ignored('exchange', 'abc'))

    def test_config_not_ignored(self):
        """
        Asserts that ignored_data is properly ignored.
        """
        ignored_data = dict(exchange=['a.*', 'b.*'])
        conf = utils.Config(self.auth, self.conn, ignored_data)
        self.assertFalse(conf.is_ignored('exchange', 'notignored'))

if __name__ == '__main__':

    logging.basicConfig(stream=sys.stderr)
    logging.getLogger().setLevel(logging.DEBUG)
    unittest.main()


travis_pypi_setup.py
#!/usr/bin/env python
# -*- coding: iso-8859-15 -*-

# Copyright (c) 2014 The New York Times Company
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Update encrypted deploy password in Travis config file
"""


from __future__ import print_function
import base64
from getpass import getpass
import json
import os
import yaml

from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.asymmetric.padding import PKCS1v15
from cryptography.hazmat.primitives.serialization import load_pem_public_key


try:
    from urllib import urlopen
except ImportError:
    from urllib.request import urlopen


GITHUB_REPO = 'NYTimes/collectd-rabbitmq'
TRAVIS_CONFIG_FILE = os.path.join(
    os.path.dirname(os.path.abspath(__file__)), '.travis.yml')


def load_key(pubkey):
    """
    Load public RSA key, with work-around for keys using
    incorrect header/footer format.

    Read more about RSA encryption with cryptography:
    https://cryptography.io/latest/hazmat/primitives/asymmetric/rsa/
    """
    try:
        return load_pem_public_key(pubkey.encode(), default_backend())
    except ValueError:
        # workaround for https://github.com/travis-ci/travis-api/issues/196
        pubkey = pubkey.replace('BEGIN RSA', 'BEGIN').replace('END RSA', 'END')
        return load_pem_public_key(pubkey.encode(), default_backend())


def encrypt(pubkey, password):
    """
    Encrypt password using given RSA public key and encode it with base64.

    The encrypted password can only be decrypted by someone with the
    private key (in this case, only Travis).
    """
    key = load_key(pubkey)
    encrypted_password = key.encrypt(password, PKCS1v15())
    return base64.b64encode(encrypted_password)


def fetch_public_key(repo):
    """
    Download RSA public key Travis will use for this repo.

    Travis API docs: http://docs.travis-ci.com/api/#repository-keys
    """
    keyurl = 'https://api.travis-ci.org/repos/{0}/key'.format(repo)
    data = json.loads(urlopen(keyurl).read().decode())
    if 'key' not in data:
        errmsg = "Could not find public key for repo: {}.\n".format(repo)
        errmsg += "Have you already added your GitHub repo to Travis?"
        raise ValueError(errmsg)
    return data['key']


def prepend_line(filepath, line):
    """
    Rewrite a file adding a line to its beginning.
    """
    with open(filepath) as f:
        lines = f.readlines()

    lines.insert(0, line)

    with open(filepath, 'w') as f:
        f.writelines(lines)


def load_yaml_config(filepath):
    """
    Returns a load YAML file from filepath
    """
    with open(filepath) as f:
        return yaml.load(f)


def save_yaml_config(filepath, config):
    with open(filepath, 'w') as f:
        yaml.dump(config, f, default_flow_style=False)


def update_travis_deploy_password(encrypted_password):
    """
    Update the deploy section of the .travis.yml file
    to use the given encrypted password.
    """
    config = load_yaml_config(TRAVIS_CONFIG_FILE)

    config['deploy']['password'] = dict(secure=encrypted_password)

    save_yaml_config(TRAVIS_CONFIG_FILE, config)

    line = ('# This file was autogenerated and will overwrite'
            ' each time you run travis_pypi_setup.py\n')
    prepend_line(TRAVIS_CONFIG_FILE, line)


def main(args):
    public_key = fetch_public_key(args.repo)
    password = args.password or getpass('PyPI password: ')
    update_travis_deploy_password(encrypt(public_key, password.encode()))
    print("Wrote encrypted password to .travis.yml -- you're ready to deploy")


if '__main__' == __name__:
    import argparse
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--repo', default=GITHUB_REPO,
                        help='GitHub repo (default: %s)' % GITHUB_REPO)
    parser.add_argument('--password',
                        help='PyPI password (will prompt if not provided)')

    args = parser.parse_args()
    main(args)


collectd_rabbitmq/utils.py
# -*- coding: iso-8859-15 -*-

# Copyright (c) 2014 The New York Times Company
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

""" Module that contains utility classes and functions """

import re
from urlparse import urlparse


class Auth(object):
    """
    Stores Auth data.
    """
    def __init__(self, username='guest', password='guest', realm=None):
        self.username = username
        self.password = password
        self.realm = realm or "RabbitMQ Management"


class ConnectionInfo(object):
    """
    Stores connection information.
    """
    def __init__(self, host='localhost', port=15672, scheme='http'):
        self.host = host
        self.port = port
        self.scheme = scheme

    @property
    def url(self):
        """
        Returns a url made from scheme, host and port.
        """
        return "{0}://{1}:{2}".format(self.scheme, self.host, self.port)

    @url.setter
    def url(self, value):
        """
        Sets scheme, host, and port from URL.
        """
        parsed_url = urlparse(value)
        self.host = parsed_url.hostname
        self.port = parsed_url.port
        self.scheme = parsed_url.scheme


class Config(object):
    """
    Class that contains configuration data.
    """
    def __init__(self, auth, connection, data_to_ignore=None):
        self.auth = auth
        self.connection = connection
        self.data_to_ignore = dict()

        if data_to_ignore:
            for key, values in data_to_ignore.items():
                self.data_to_ignore[key] = list()
                for value in values:
                    self.data_to_ignore[key].append(re.compile(value))

    def is_ignored(self, stat_type, name):
        """
        Return true if name of type qtype should be ignored.
        """
        if stat_type in self.data_to_ignore:
            for regex in self.data_to_ignore[stat_type]:
                match = regex.match(name)
                if match:
                    return True
        return False


def filter_dictionary(dictionary, keys):
    """
    Returns a dictionary with only keys.
    """
    if not keys:
        return dict()

    if not dictionary:
        return dict()

    return dict((key, dictionary[key]) for key in keys if key in dictionary)


def is_sequence(arg):
    """
    Returns true if arg behaves like a sequence,
    unless it also implements strip, such as strings.
    """

    return (not hasattr(arg, "strip") and
            hasattr(arg, "__getitem__") or
            hasattr(arg, "__iter__"))


collectd_rabbitmq/rabbit.py
# -*- coding: iso-8859-15 -*-

# Copyright (c) 2014 The New York Times Company
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
python plugin for collectd to obtain rabbitmq stats
"""

import collectd
import json
import urllib
import urllib2


class RabbitMQStats(object):
    """
        Class to interface with the RabbitMQ API.
    """
    def __init__(self, config):
        self.config = config
        self.api = "{0}/api".format(self.config.connection.url)

    @staticmethod
    def get_names(items):
        """
        Return URL encoded names.
        """
        collectd.debug("Getting names for %s" % items)
        names = list()
        for item in items:
            name = item.get('name', None)
            if name:
                name = urllib.quote(name, '')
                names.append(name)
        return names

    def get_info(self, *args):
        """
        return JSON object from URL.
        """

        url = "{0}/{1}".format(self.api, '/'.join(args))
        collectd.debug("Getting info for %s" % url)

        auth_handler = urllib2.HTTPBasicAuthHandler()
        auth_handler.add_password(realm=self.config.auth.realm,
                                  uri=self.api,
                                  user=self.config.auth.username,
                                  passwd=self.config.auth.password)
        opener = urllib2.build_opener(auth_handler)
        urllib2.install_opener(opener)

        try:
            info = urllib2.urlopen(url)
        except urllib2.HTTPError as http_error:
            collectd.error("HTTP Error: %s" % http_error)
            return None
        except urllib2.URLError as url_error:
            collectd.error("URL Error: %s" % url_error)
            return None
        except ValueError as value_error:
            collectd.error("Value Error: %s" % value_error)
            return None

        try:
            return_value = json.load(info)
        except ValueError as err:
            collectd.error("ValueError parsing JSON from %s: %s" % (url, err))
            return_value = None
        except TypeError as err:
            collectd.error("TypeError parsing JSON from %s: %s" % (url, err))
            return_value = None
        return return_value

    def get_nodes(self):
        """
        Return a list of nodes.
        """
        return self.get_info("nodes") or list()
    nodes = property(get_nodes)

    # Vhosts
    def get_vhosts(self):
        """
        Returns a list of vhosts.
        """
        collectd.debug("Getting a list of vhosts")
        return self.get_info("vhosts") or list()

    def get_vhost_names(self):
        """
        Returns a list of vhost names.
        """
        collectd.debug("Getting vhost names")
        all_vhosts = self.get_vhosts()
        return self.get_names(all_vhosts) or list()
    vhost_names = property(get_vhost_names)

    # Exchanges
    def get_exchanges(self, vhost_name=None):
        """
        Returns raw exchange data.
        """
        collectd.debug("Getting exchanges for %s" % vhost_name)
        return self.get_info("exchanges", vhost_name)

    def get_exchange_names(self, vhost_name=None):
        """
        Returns a list of all exchange names.
        """
        collectd.debug("Getting exchange names for %s" % vhost_name)
        all_exchanges = self.get_exchanges(vhost_name)
        return self.get_names(all_exchanges)

    def get_exchange_stats(self, exchange_name=None, vhost_name=None):
        """
        Returns a dictionary of stats for exchange_name.
        """
        collectd.debug("Getting exchange stats for %s in %s" %
                       (exchange_name, vhost_name))

        return self.get_stats('exchange', exchange_name, vhost_name)

    # Queues
    def get_queues(self, vhost_name=None):
        """
        Returns raw queue data.
        """
        collectd.debug("Getting queues for %s" % vhost_name)
        return self.get_info("queues", vhost_name)

    def get_queue_names(self, vhost_name=None):
        """
        Returns a list of all queue names.
        """
        collectd.debug("Getting queue names for %s" % vhost_name)
        all_queues = self.get_queues(vhost_name)
        return self.get_names(all_queues)

    def get_queue_stats(self, queue_name=None, vhost_name=None):
        """
        Returns a dictionary of stats for queue_name.
        """
        return self.get_stats('queue', queue_name, vhost_name)

    def get_stats(self, stat_type, stat_name, vhost_name):
        """
        Returns a dictionary of stats.
        """
        collectd.debug("Getting stats for %s %s%s in %s" %
                       (stat_name or 'all',
                        stat_type,
                        's' if not stat_name else '',
                        vhost_name))

        if stat_type not in('exchange', 'queue'):
            raise ValueError("Unsupported stat type {0}".format(stat_type))
        stat_name_func = getattr(self, 'get_{0}_names'.format(stat_type))
        if not vhost_name:
            vhosts = self.get_vhost_names()
        else:
            vhosts = [vhost_name]

        stats = dict()
        for vhost in vhosts:
            if not stat_name:
                names = stat_name_func(vhost)
            else:
                names = [stat_name]
            for name in names:
                if not self.config.is_ignored(stat_type, name):
                    stats[name] = self.get_info("{0}s".format(stat_type),
                                                vhost,
                                                name)
        return stats


rabbitmq.py
"""
python plugin for collectd to obtain rabbitmq stats
"""
import collectd
import urllib2
import urllib
import json
import re

RABBIT_API_URL = "http://{host}:{port}/api/"

QUEUE_MESSAGE_STATS = ['messages', 'messages_ready', 'messages_unacknowledged']
QUEUE_STATS = ['memory', 'messages', 'consumers']

MESSAGE_STATS = ['ack', 'publish', 'publish_in', 'publish_out', 'confirm',
                 'deliver', 'deliver_noack', 'get', 'get_noack', 'deliver_get',
                 'redeliver', 'return']
MESSAGE_DETAIL = ['avg', 'avg_rate', 'rate', 'sample']
NODE_STATS = ['disk_free', 'disk_free_limit', 'fd_total',
              'fd_used', 'mem_limit', 'mem_used',
              'proc_total', 'proc_used', 'processors', 'run_queue',
              'sockets_total', 'sockets_used']

PLUGIN_CONFIG = {
    'username': 'guest',
    'password': 'guest',
    'host': 'localhost',
    'port': 15672,
    'realm': 'RabbitMQ Management'
}


def configure(config_values):
    '''
    Load information from configuration file
    '''

    global PLUGIN_CONFIG
    collectd.info('Configuring RabbitMQ Plugin')
    for config_value in config_values.children:
        collectd.info("%s = %s" % (config_value.key,
                                   len(config_value.values) > 0))
        if len(config_value.values) > 0:
            if config_value.key == 'Username':
                PLUGIN_CONFIG['username'] = config_value.values[0]
            elif config_value.key == 'Password':
                PLUGIN_CONFIG['password'] = config_value.values[0]
            elif config_value.key == 'Host':
                PLUGIN_CONFIG['host'] = config_value.values[0]
            elif config_value.key == 'Port':
                PLUGIN_CONFIG['port'] = config_value.values[0]
            elif config_value.key == 'Realm':
                PLUGIN_CONFIG['realm'] = config_value.values[0]
            elif config_value.key == 'Ignore':
                type_rmq = config_value.values[0]
                PLUGIN_CONFIG.setdefault('ignore', {})[type_rmq] = []
                for regex in config_value.children:
                    PLUGIN_CONFIG['ignore'][type_rmq].append(
                        re.compile(regex.values[0]))


def init():
    '''
    Initalize connection to rabbitmq
    '''
    collectd.info('Initalizing RabbitMQ Plugin')


def get_info(url):
    '''
    return json object from url
    '''

    try:
        info = urllib2.urlopen(url)
    except urllib2.HTTPError as http_error:
        collectd.error("Error: %s" % (http_error))
        return None
    except urllib2.URLError as url_error:
        collectd.error("Error: %s" % (url_error))
        return None
    return json.load(info)


def dispatch_values(values, host, plugin, plugin_instance, metric_type,
                    type_instance=None):
    '''
    dispatch metrics to collectd
    Args:
      values (tuple): the values to dispatch
      host: (str): the name of the vhost
      plugin (str): the name of the plugin. Should be queue/exchange
      plugin_instance (str): the queue/exchange name
      metric_type: (str): the name of metric
      type_instance: Optional
    '''

    collectd.debug("Dispatching %s %s %s %s %s\n\t%s " % (host, plugin,
                   plugin_instance, metric_type, type_instance, values))

    metric = collectd.Values()
    if host:
        metric.host = host
    metric.plugin = plugin
    if plugin_instance:
        metric.plugin_instance = plugin_instance
    metric.type = metric_type
    if type_instance:
        metric.type_instance = type_instance
    metric.values = values
    # Tiny hack to fix bug with write_http plugin in Collectd versions < 5.5.
    # See https://github.com/phobos182/collectd-elasticsearch/issues/15 for details
    metric.meta = {'0': True}
    metric.dispatch()


def dispatch_message_stats(data, vhost, plugin, plugin_instance):
    """
    Sends message stats to collectd.
    """
    if not data:
        collectd.debug("No data for %s in vhost %s" % (plugin, vhost))
        return

    for name in MESSAGE_STATS:
        dispatch_values((data.get(name, 0),), vhost, plugin,
                        plugin_instance, name)


def generate_vhost_name(vhost):
    '''
    Generate a "normalized" vhost name without /
    '''
    name = vhost['name']
    if name == '/':
        name = 'default'
    else:
        name = re.sub(r'^/', 'slash_', name)
        name = re.sub(r'/$', '_slash', name)
        name = re.sub(r'/', '_slash_', name)
    return 'rabbitmq_%s' % name


def dispatch_queue_metrics(queue, vhost):
    '''
    Dispatches queue metrics for queue in vhost
    '''

    vhost_name = generate_vhost_name(vhost)
    for name in QUEUE_STATS:
        values = list((queue.get(name, 0),))
        dispatch_values(values, vhost_name, 'queues', queue['name'],
                        'rabbitmq_%s' % name)

    for name in QUEUE_MESSAGE_STATS:
        values = list((queue.get(name, 0),))
        dispatch_values(values, vhost_name, 'queues', queue['name'],
                        'rabbitmq_%s' % name)

        details = queue.get("%s_details" % name, None)
        if not details:
            continue
        values = list()
        for detail in MESSAGE_DETAIL:
            values.append(details.get(detail, 0))
        dispatch_values(values, vhost_name, 'queues', queue['name'],
                        'rabbitmq_details', name)

    dispatch_message_stats(queue.get('message_stats', None), vhost_name,
                           'queues', queue['name'])


def dispatch_exchange_metrics(exchange, vhost):
    '''
    Dispatches exchange metrics for exchange in vhost
    '''
    vhost_name = generate_vhost_name(vhost)
    dispatch_message_stats(exchange.get('message_stats', None), vhost_name,
                           'exchanges', exchange['name'])


def dispatch_node_metrics(node):
    '''
    Dispatches node metrics
    '''

    for name in NODE_STATS:
        dispatch_values((node.get(name, 0),), node['name'].split('@')[1],
                        'rabbitmq', None, name)


def want_to_ignore(type_rmq, name):
    """
    Applies ignore regex to the queue.
    """
    if 'ignore' in PLUGIN_CONFIG:
        if type_rmq in PLUGIN_CONFIG['ignore']:
            for regex in PLUGIN_CONFIG['ignore'][type_rmq]:
                match = regex.match(name)
                if match:
                    return True
    return False


def read(input_data=None):
    '''
    reads all metrics from rabbitmq
    '''

    collectd.debug("Reading data with input = %s" % (input_data))
    base_url = RABBIT_API_URL.format(host=PLUGIN_CONFIG['host'],
                                     port=PLUGIN_CONFIG['port'])

    auth_handler = urllib2.HTTPBasicAuthHandler()
    auth_handler.add_password(realm=PLUGIN_CONFIG['realm'],
                              uri=base_url,
                              user=PLUGIN_CONFIG['username'],
                              passwd=PLUGIN_CONFIG['password'])
    opener = urllib2.build_opener(auth_handler)
    urllib2.install_opener(opener)

    #First get all the nodes
    for node in get_info("%s/nodes" % (base_url)):
        dispatch_node_metrics(node)

    #Then get all vhost

    for vhost in get_info("%s/vhosts" % (base_url)):

        vhost_name = urllib.quote(vhost['name'], '')
        collectd.debug("Found vhost %s" % vhost['name'])

        for queue in get_info("%s/queues/%s" % (base_url, vhost_name)):
            queue_name = urllib.quote(queue['name'], '')
            collectd.debug("Found queue %s" % queue['name'])
            if not want_to_ignore("queue", queue_name):
                queue_data = get_info("%s/queues/%s/%s" % (base_url,
                                                           vhost_name,
                                                           queue_name))
                if queue_data is not None:
                    dispatch_queue_metrics(queue_data, vhost)
                else:
                    collectd.warning("Cannot get data back from %s/%s queue" %
                                    (vhost_name, queue_name))

        for exchange in get_info("%s/exchanges/%s" % (base_url,
                                 vhost_name)):
            exchange_name = urllib.quote(exchange['name'], '')
            if exchange_name:
                collectd.debug("Found exchange %s" % exchange['name'])
                if not want_to_ignore("exchange", queue_name):
                    exchange_data = get_info("%s/exchanges/%s/%s" % (
                                             base_url, vhost_name, exchange_name))
                    dispatch_exchange_metrics(exchange_data, vhost)


def shutdown():
    '''
    Shutdown connection to rabbitmq
    '''

    collectd.info('RabbitMQ plugin shutting down')

# Register callbacks
collectd.register_config(configure)
collectd.register_init(init)
collectd.register_read(read)
#collectd.register_write(write)
collectd.register_shutdown(shutdown)
